tests/test_cases/test_cocotb/test_scheduler.py

```diff
@@ -858,3 +858,83 @@ async def noop():
     j = await Join(t)
     assert isinstance(j, Join)
     assert j.task is t
+
+
+@cocotb.test
+async def test_task_done_callback_passing(_) -> None:
+    callback_ran = False
+
+    def done_callback(_: Task) -> None:
+        nonlocal callback_ran
+        callback_ran = True
+
+    async def passing_coro() -> None:
+        pass
+
+    passing_task = cocotb.start_soon(passing_coro())
+    passing_task.add_done_callback(done_callback)
+    await passing_task
+    assert callback_ran
+
+
+@cocotb.test
+async def test_task_done_callback_erroring(_) -> None:
+    callback_ran = False
+
+    def done_callback(_: Task) -> None:
+        nonlocal callback_ran
+        callback_ran = True
+
+    async def erroring_coro() -> None:
+        raise Exception
+
+    erroring_task = cocotb.start_soon(erroring_coro())
+    callback_ran = False
+    erroring_task.add_done_callback(done_callback)
+    try:
+        await erroring_task
+    except Exception:
+        pass
+    assert callback_ran
+
+
+@cocotb.test
+async def test_task_done_callback_cancelled(_) -> None:
+    callback_ran = False
+
+    def done_callback(_: Task) -> None:
+        nonlocal callback_ran
+        callback_ran = True
+
+    async def cancelled_coro() -> None:
+        e = Event()
+        await e.wait()
+
+    cancelled_task = cocotb.start_soon(cancelled_coro())
+    callback_ran = False
+    cancelled_task.add_done_callback(done_callback)
+    await Timer(1, "ns")
+    with pytest.warns(FutureWarning):
+        cancelled_task.cancel()
+    await NullTrigger()
+    assert callback_ran
+
+
+@cocotb.test
+async def test_task_done_callback_removed(_) -> None:
+    callback_ran = False
+
+    def done_callback(_: Task) -> None:
+        nonlocal callback_ran
+        callback_ran = True
+
+    async def passing_coro() -> None:
+        await Timer(10, "ns")
+
+    passing_task = cocotb.start_soon(passing_coro())
+    passing_task.add_done_callback(done_callback)
+    await Timer(1, "ns")
+    assert not callback_ran
+    passing_task.remove_done_callback(done_callback)
+    await passing_task
+    assert not callback_ran
```

Comment:
Add another call `passing_task.remove_done_callback(done_callback)` that should return `0`.